File: java/src/main/java/sa/amnn/kit/resources/Customers.java
package sa.amnn.kit.resources;

import com.fasterxml.jackson.core.type.TypeReference;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;
import sa.amnn.kit.AmenClient;
import sa.amnn.kit.Models.CreateCustomer;
import sa.amnn.kit.Models.Customer;
import sa.amnn.kit.Models.Page;

public final class Customers {
    private final AmenClient c; public Customers(AmenClient c) { this.c = c; }
    public Customer create(CreateCustomer body) { return c.post("/customers/", body, new TypeReference<>() {}); }
    public Customer get(String customerNumber) { return c.get("/customers/" + customerNumber, Map.of(), new TypeReference<>() {}); }
    public Page<Customer> list(Map<String, String> params) { return c.page("/customers/", params, "customers", Customer.class); }
    /** Iterate every page — never process only the first page by accident. */
    public Iterable<Customer> all(Map<String, String> filters) {
        return () -> new Iterator<Customer>() {
            int page = 0; Page<Customer> cur = fetch(); int i = 0;
            private Page<Customer> fetch() { Map<String, String> p = new HashMap<>(filters); p.put("page", String.valueOf(page)); return list(p); }
            public boolean hasNext() { if (i < cur.items().size()) return true; if (page + 1 >= cur.pages() || cur.items().isEmpty()) return false; page++; cur = fetch(); i = 0; return !cur.items().isEmpty(); }
            public Customer next() { return cur.items().get(i++); }
        };
    }
}

File: kotlin/amen-client/src/main/kotlin/sa/amnn/kit/resources/Customers.kt
package sa.amnn.kit.resources

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.serialization.json.encodeToJsonElement
import sa.amnn.kit.*

class Customers(private val c: AmenClient) {
    suspend fun create(body: CreateCustomer) = c.request("POST", "/customers/", Customer.serializer(), body = json.encodeToJsonElement(body))!!
    suspend fun get(customerNumber: String) = c.request("GET", "/customers/$customerNumber", Customer.serializer())!!
    suspend fun list(params: Map<String, String?> = emptyMap()) = c.page("/customers/", params, "customers", Customer.serializer())
    /** Iterate every page — never process only the first page by accident. */
    fun all(filters: Map<String, String?> = emptyMap()): Flow<Customer> = flow {
        var page = 0
        while (true) {
            val p = list(filters + ("page" to page.toString()))
            p.items.forEach { emit(it) }
            if (page + 1 >= p.pages || p.items.isEmpty()) return@flow
            page++
        }
    }
}
